src/CommunityToolkit/Xamarin.CommunityToolkit/Views/Calendar/Calendar.shared.cs

```diff
@@ -0,0 +1,533 @@
+		{
+			UpdateGridForDays();
+			UpdateDays();
+			SelectDays(SelectedDays);
+		}
+
+		void SelectDays(IEnumerable<DateTime> daysSelected)
+		{
+			foreach (var day in days)
+			{
+				var isSelected = daysSelected.Any(x => x.Date == day.Date);
+				day.IsSelected = isSelected;
+			}
+		}
+
+		void UpdateGridForDays()
+		{
+			if (!ShowWeekends)
+			{
+				gridDays.ColumnDefinitions[5].Width = new GridLength(0);
+				gridDays.ColumnDefinitions[6].Width = new GridLength(0);
+			}
+			else if (ShowWeekends)
+			{
+				gridDays.ColumnDefinitions[5].Width = GridLength.Star;
+				gridDays.ColumnDefinitions[6].Width = GridLength.Star;
+			}
+
+			var weeksInMonth = Date.WeeksInMonth(FirstDayOfWeek);
+
+			if (gridDays.RowDefinitions.Count < weeksInMonth)
+			{
+				for (var row = gridDays.RowDefinitions.Count; row < weeksInMonth; row++)
+				{
+					gridDays.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto });
+
+					var daysToAdd = days.Where(x => Grid.GetRow(x) == row).ToList();
+
+					foreach (var calendarDay in daysToAdd)
+					{
+						gridDays.Children.Add(calendarDay);
+					}
+				}
+			}
+			else if (gridDays.RowDefinitions.Count > weeksInMonth)
+			{
+				for (var row = gridDays.RowDefinitions.Count; row > weeksInMonth; row--)
+				{
+					gridDays.RowDefinitions.RemoveAt(gridDays.RowDefinitions.Count - 1);
+
+					var daysToRemove = days.Where(x => Grid.GetRow(x) == row - 1).ToList();
+
+					foreach (var calendarDay in daysToRemove)
+					{
+						gridDays.Children.Remove(calendarDay);
+					}
+				}
+			}
+		}
+
+		void UpdateDays()
+		{
+			var weeksInMonth = Date.WeeksInMonth(FirstDayOfWeek);
+			var daysInMonth = Date.DaysInMonth();
+
+			for (var day = 1; day <= daysInMonth; day++)
+			{
+				var date = new DateTime(Date.Year, Date.Month, day);
+				var weekOfMonth = date.WeekOfMonth(FirstDayOfWeek);
+
+				UpdateDaysFromPreviousOrNextMonths(weekOfMonth, date, weeksInMonth, daysInMonth);
+
+				if (!ShowWeekends && (date.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+				{
+					continue;
+				}
+
+				UpdateDay(date, weekOfMonth, isVisible: true);
+			}
+
+			if (SelectedDays?.Count > 0)
+			{
+				SelectDays(SelectedDays);
+			}
+		}
+
+		void UpdateDaysFromPreviousOrNextMonths(
+			int week,
+			DateTime date,
+			int weeksInMonth,
+			int daysInMonth)
+		{
+			if (week is 1
+				&& date.Day is 1
+				&& date.DayOfWeek != FirstDayOfWeek)
+			{
+				var newDate = date;
+
+				do
+				{
+					newDate = newDate.AddDays(-1);
+
+					if (!ShowWeekends && (newDate.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday))
+					{
+						continue;
+					}
+
+					UpdateDay(newDate, week, isVisible: ShowDaysFromOtherMonths);
+				} while (newDate.DayOfWeek != FirstDayOfWeek);
+			}
+			else if (week == weeksInMonth
+					 && date.Day == daysInMonth
+					 && date.DayOfWeek != FirstDayOfWeek.PreviousOrFirst())
+			{
+				var newDate = date;
+				var lastDayOfWeek = FirstDayOfWeek.PreviousOrFirst();
```

Comment:
can we cache the result for `FirstDayOfWeek.PreviousOrFirst()` avoiding duplicate calls?
```suggestion
                    // somewhere at the top
                    var firstDayOfWeek = FirstDayOfWeek.PreviousOrFirst();
			else if (week == weeksInMonth
					 && date.Day == daysInMonth
					 && date.DayOfWeek != firstDayOfWeek)
			{
				var newDate = date;
				var lastDayOfWeek = firstDayOfWeek;
```

Reply:
value is now being cached at the top of the method